### PR TITLE
formatter: support for virtual host metadata

### PR DESCRIPTION
**Pull request by mathetake**
Commit Message: formatter: support for virtual host metadata
Additional Description:
This enables the access to the virtual host metadata introduced in https://github.com/envoyproxy/envoy/pull/30175 from formatter.

Risk Level: low 
Testing: : unit test
Docs Changes: done.
Release Notes:  
Platform Specific Features:
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/34900

### CODE REVIEW

**Comment by mathetake** [comment]
Verify Examples job failure is irrelevant to this eg the same error on main https://github.com/envoyproxy/envoy/actions/runs/9708036595/job/26794181496

**Review by abeyad — DISMISSED** [review]
/lgtm api

**Comment by mathetake** [comment]
thanks @RyanTheOptimist, added an integration test with access logger in https://github.com/envoyproxy/envoy/pull/34958/commits/747b64282442575299f39e4f53e7848f7f5832a3

**Comment by RyanTheOptimist** [comment]
> thanks @RyanTheOptimist, added an integration test with access logger in [747b642](https://github.com/envoyproxy/envoy/commit/747b64282442575299f39e4f53e7848f7f5832a3)

This is a good test, but I was hoping to see something in `test/integration` in which we see Envoy process an actual request. This test is TCP-proxy specific, but shows the idea. https://github.com/envoyproxy/envoy/blob/main/test/integration/tcp_proxy_integration_test.cc#L378

Would this be possible?

**Comment by mathetake** [comment]
well this doesn't work as well - 
```
vscode@mathetake-Venus-series:/workspaces/envoy$ bazel test --define wasm=none //test/integration:tcp_proxy_integration_test
ERROR: /workspaces/envoy/test/integration/BUILD:1750:24: in cc_test rule //test/integration:tcp_proxy_integration_test: target '//source/extensions/formatter/metadata:config' is not visible from target '//test/integration:tcp_proxy_integration_test'. Check the visibility declaration of the former target if you think the dependency is legitimate
ERROR: /workspaces/envoy/test/integration/BUILD:1750:24: Analysis of target '//test/integration:tcp_proxy_integration_test' failed
ERROR: Analysis of target '//test/integration:tcp_proxy_integration_test' failed; build aborted: 
INFO: Elapsed time: 0.938s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (605 packages loaded, 35042 targets configured)
ERROR: Couldn't start the build. Unable to run tests
```

**Comment by RyanTheOptimist** [comment]
> so
> 
> ```diff
> diff --git a/source/extensions/formatter/metadata/BUILD b/source/extensions/formatter/metadata/BUILD
> index 8d1bf6bf39..78a8e7d86b 100644
> --- a/source/extensions/formatter/metadata/BUILD
> +++ b/source/extensions/formatter/metadata/BUILD
> @@ -23,6 +23,10 @@ envoy_cc_extension(
>      name = "config",
>      srcs = ["config.cc"],
>      hdrs = ["config.h"],
> +    extra_visibility = [
> +        "//test:__subpackages__",
> +    ],
>      deps = [
>          "//envoy/registry",
>          "//source/extensions/formatter/metadata:metadata_lib",
> diff --git a/test/integration/BUILD b/test/integration/BUILD
> index 5f643fb68c..d5dd5ebc54 100644
> --- a/test/integration/BUILD
> +++ b/test/integration/BUILD
> @@ -1767,6 +1767,7 @@ envoy_cc_test(
>          "//source/extensions/access_loggers/file:config",
>          "//source/extensions/filters/network/common:factory_base_lib",
>          "//source/extensions/filters/network/tcp_proxy:config",
> +        "//source/extensions/formatter/metadata:config",
>          "//source/extensions/load_balancing_policies/subset:config",
>          "//source/extensions/transport_sockets/tls:config",
>          "//test/mocks/runtime:runtime_mocks",
> ```
> 
> ok so this is what we need. is this acceptable? @RyanTheOptimist

Hm. Good question. I have a vague recollection that we have a policy about this somehow. @alyssawilk can you clarify?

**Comment by alyssawilk** [comment]
generally we request folks not add tests of extension functionality into "core" directories but instead have e2e tests in the extension directory.  Otherwise we often break the build for folks who don't import a given extension

**Comment by mathetake** [comment]
thank you for clarifying. So I think my last commit https://github.com/envoyproxy/envoy/commit/747b64282442575299f39e4f53e7848f7f5832a3 is the furthest we can get without adding the extra visibility.

**Comment by RyanTheOptimist** [comment]
> thank you for clarifying. So I think my last commit [747b642](https://github.com/envoyproxy/envoy/commit/747b64282442575299f39e4f53e7848f7f5832a3) is the furthest we can get without adding the extra visibility.
> 
> edited: we could add `test/extensions/formatter/metadata/integration_test.cc` in the first place, but that would go beyond the scope of this PR since I have to write tests for all metadata types, not only for VIRTUAL_HOST that is added in this PR. I think I would do that in a follow-up PR later. wdyt? @RyanTheOptimist

Doing this in a follow up would be fine.

**Comment by RyanTheOptimist** [comment]
/wait
on CI

**Comment by mathetake** [comment]
the verify test failing looks irrelevant to this PR.

File: source/extensions/formatter/metadata/metadata.cc [hunk 1]
```diff
@@ -46,6 +46,24 @@ class ListenerMetadataFormatter : public ::Envoy::Formatter::MetadataFormatter {
...
+                               absl::optional<size_t> max_length)
+      : ::Envoy::Formatter::MetadataFormatter(
+            filter_namespace, path, max_length,
+            [](const StreamInfo::StreamInfo& stream_info)
+                -> const envoy::config::core::v3::Metadata* {
+              Router::RouteConstSharedPtr route = stream_info.route();
+              if (route == nullptr) {
+                return nullptr;
+              }
+              return &route->routeEntry()->virtualHost().metadata();
```
Comment (wbpcode) [inline source/extensions/formatter/metadata/metadata.cc:63]: Note, there is no guarantee that the route always has a valid route entry.

File: test/extensions/access_loggers/file/BUILD [hunk 2]
```diff
@@ -14,9 +14,13 @@ envoy_package()
...
     name = "config_test",
     srcs = ["config_test.cc"],
-    extension_names = ["envoy.access_loggers.file"],
+    extension_names = [
+        "envoy.access_loggers.file",
+        "envoy.formatter.metadata",
+    ],
     deps = [
         "//source/extensions/access_loggers/file:config",
+        "//source/extensions/formatter/metadata:config",
```
Comment (wbpcode) [inline test/extensions/access_loggers/file/BUILD:23]: Please don't do this. Basically, we will use the `extension_build_config.bzl` to control whether an extension should be linked or not in the final binary.

This change introduce unnecessary dependency and break the above rule.

File: test/extensions/access_loggers/file/config_test.cc [hunk 3]
```diff
@@ -176,6 +176,33 @@ TEST_F(FileAccessLogTest, LogFormatJson) {
...
+  log_format:
+    formatters:
+    - name: envoy.formatter.metadata
+      typed_config:
+        "@type": type.googleapis.com/envoy.extensions.formatter.metadata.v3.Metadata
+    text_format_source:
+      inline_string: "%REQ(:path)% metadata=%METADATA(VIRTUAL_HOST:metadata.namespace:envoy)%"
+)",
+      "/bar/foo metadata=proxy", false);
+}
```
Comment (wbpcode) [inline test/extensions/access_loggers/file/config_test.cc:204]: Although more tests are better, I think this test is redundant if you test this new feature in the metadata_test.cc carefully.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for the contribution. Some comments are added.

**Comment by wbpcode** [comment]
> @phlax verify_examples wasm-cc is constantly failing all over the places recently. Anyone is working on fixing it?

This won't block this work anyway.

File: test/extensions/access_loggers/file/config_test.cc (hunk 3 above)
Comment (mathetake) [reply test/extensions/access_loggers/file/config_test.cc]: https://github.com/envoyproxy/envoy/pull/34958/commits/fbe99e13e36ab5817aac69d2d660a69702e2ba5a reverted the test

File: source/extensions/formatter/metadata/metadata.cc (hunk 1 above)
Comment (mathetake) [reply source/extensions/formatter/metadata/metadata.cc]: done handling the invalid route entry https://github.com/envoyproxy/envoy/pull/34958/commits/153fc582e3d70e1d7dbba8b0e82d91c9634cb7a1

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
@wbpcode friendly ping

**Review by wbpcode — APPROVED** [review]

**Pull request merged**